File: lab09_stack/youbotKDL_tester/src/YoubotKDL.cpp
#include <youbotKDL_tester/YoubotKDL.h>

void YoubotKDL::init()
{
    DH_params[0][0] = -0.033;   DH_params[1][0] = 0.155;  DH_params[2][0] = 0.135; DH_params[3][0] = -0.002;  DH_params[4][0] = 0.0;
    DH_params[0][1] = M_PI_2;   DH_params[1][1] = 0.0;    DH_params[2][1] = 0.0;   DH_params[3][1] = M_PI_2;  DH_params[4][1] = M_PI;
    DH_params[0][2] = 0.145;    DH_params[1][2] = 0.0;    DH_params[2][2] = 0.0;   DH_params[3][2] = 0.0;     DH_params[4][2] = -0.185;
    DH_params[0][3] = M_PI;     DH_params[1][3] = M_PI_2; DH_params[2][3] = 0.0;   DH_params[3][3] = -M_PI_2; DH_params[4][3] = -M_PI;

    DH_params[0][3] = 170*M_PI/180;
    DH_params[1][3] = 65*M_PI/180+M_PI_2;
    DH_params[2][3] = -146*M_PI/180;
    DH_params[3][3] = M_PI_2+102.5*M_PI/180;
    DH_params[4][3] = M_PI+167.5*M_PI/180;
    DH_params[0][4] = DH_params[1][4] = DH_params[2][4] = DH_params[3][4] = DH_params[4][4] = -1;

    joint_offset[0] = 170*M_PI/180;
    joint_offset[1] = 65*M_PI/180;
    joint_offset[2] = -146*M_PI/180;
    joint_offset[3] = 102.5*M_PI/180;
    joint_offset[4] = 167.5*M_PI/180;

    joint_limit_min[0] = -169*M_PI/180;
    joint_limit_min[1] = -65*M_PI/180;
    joint_limit_min[2] = -151*M_PI/180;
    joint_limit_min[3] = -102.5*M_PI/180;
    joint_limit_min[4] = -167.5*M_PI/180;

    joint_limit_max[0] = 169*M_PI/180;
    joint_limit_max[1] = 90*M_PI/180;
    joint_limit_max[2] = 146*M_PI/180;
    joint_limit_max[3] = 102.5*M_PI/180;
    joint_limit_max[4] = 167.5*M_PI/180;

    this->subscriber_joint_state = this->n.subscribe<sensor_msgs::JointState>("/joint_states", 5, &YoubotKDL::joint_state_callback,
                                                                              this);
    setup_kdl_chain();
    this->current_joint_position = KDL::JntArray(this->kine_chain.getNrOfJoints());
}

KDL::Jacobian YoubotKDL::get_jacobian(KDL::JntArray current_joint_position)
{
    KDL::ChainJntToJacSolver jac_solver = KDL::ChainJntToJacSolver(this->kine_chain);
    KDL::Jacobian jac = KDL::Jacobian(this->kine_chain.getNrOfJoints());
    jac_solver.JntToJac(current_joint_position, jac);
    return jac;
}

KDL::Frame YoubotKDL::forward_kinematics(KDL::JntArray current_joint_position, KDL::Frame current_pose)
{
    KDL::ChainFkSolverPos_recursive fk_solver = KDL::ChainFkSolverPos_recursive(this->kine_chain);
    fk_solver.JntToCart(current_joint_position, current_pose, 5);
    return current_pose;
}

void YoubotKDL::joint_state_callback(const sensor_msgs::JointState::ConstPtr &q)
{
    for (int i = 0; i < 5; i++)
        this->current_joint_position.data(i) = q->position.at(i);
}

void YoubotKDL::setup_kdl_chain() {
    for (int i = 0; i < 5; i++)
        this->kine_chain.addSegment(KDL::Segment(KDL::Joint(KDL::Vector(), KDL::Vector(0, 0, DH_params[i][4]), KDL::Joint::RotAxis),
                                                 KDL::Frame::DH(DH_params[i][0], DH_params[i][1], DH_params[i][2],
                                                                DH_params[i][3])));
}

KDL::JntArray YoubotKDL::inverse_kinematics_closed(KDL::Frame desired_pose)
{

    KDL::ChainIkSolverPos_LMA ik_solver = KDL::ChainIkSolverPos_LMA(this->kine_chain);
    KDL::JntArray required_joint = KDL::JntArray(this->kine_chain.getNrOfJoints());
    ik_solver.CartToJnt(this->current_joint_position, desired_pose, required_joint);

    return required_joint;
}

void YoubotKDL::broadcast_pose(KDL::Frame current_pose)
{
    geometry_msgs::TransformStamped trans;

    trans = tf2::kdlToTransform(current_pose);
    trans.header.stamp = ros::Time::now();
    trans.header.frame_id = "base_link";
    trans.child_frame_id = "arm_end_effector";

    this->pose_broadcaster.sendTransform(trans);

}

File: lab07/youbotkdl_tester/src/youbotkdl_tester/YoubotKDL.py
#!/usr/bin/env python

import rospy
from math import pi
from sensor_msgs.msg import JointState
from geometry_msgs.msg import TransformStamped, Vector3, Quaternion
import PyKDL
import tf2_ros
from trajectory_msgs.msg import JointTrajectory, JointTrajectoryPoint

from tf_conversions import posemath

import tf2_kdl


class YoubotKDL(object):
    def __init__(self):

        self.dh_params = [[-0.033, pi / 2, 0.145, pi, -1],
                          [0.155, 0, 0, pi / 2, -1],
                          [0.135, 0, 0, 0.0, -1],
                          [-0.002, pi / 2, 0, -pi / 2, -1],
                          [0, pi, -0.185, -pi, -1]]

        self.joint_offset = [170*pi/180, 65*pi/180, -146*pi/180, 102.5*pi/180, 167.5*pi/180]

        self.joint_limit_min = [-169*pi/180, -65*pi/180, -151*pi/180, -102.5*pi/180, -167.5*pi/180]
        self.joint_limit_max = [169*pi/180, 90*pi/180, 146*pi/180, 102.5*pi/180, 167.5*pi/180]

        # Setup the subscribers for the joint states
        self.subscriber_joint_state_ = rospy.Subscriber('/joint_states', JointState, self.joint_state_callback,
                                                        queue_size=5)
        # TF2 broadcaster
        self.pose_broadcaster = tf2_ros.TransformBroadcaster()

        # PyKDL
        self.kine_chain = PyKDL.Chain()
        self.setup_kdl_chain()
        self.current_joint_position = PyKDL.JntArray(self.kine_chain.getNrOfJoints())
        self.current_pose = PyKDL.Frame()

        self.fk_solver = PyKDL.ChainFkSolverPos_recursive(self.kine_chain)
        self.ik_solver = PyKDL.ChainIkSolverPos_LMA(self.kine_chain)
        self.jac_calc = PyKDL.ChainJntToJacSolver(self.kine_chain)

    def setup_kdl_chain(self):
        for dh in self.dh_params:
            self.kine_chain.addSegment(PyKDL.Segment(PyKDL.Joint(PyKDL.Vector(), PyKDL.Vector(0, 0, dh[4]),
                                                                 PyKDL.Joint.RotAxis),
                                                     PyKDL.Frame().DH(dh[0], dh[1], dh[2], dh[3])))

    def joint_state_callback(self, msg):
        # Copies joint position into KDL JntArray
        for i in range(0, 5):
            self.current_joint_position[i] = msg.position[i]

    def forward_kinematics(self, joint, pose):
        self.fk_solver.JntToCart(joint, pose)

    def broadcast_pose(self, pose):
        trans = TransformStamped()

        t = posemath.toMsg(pose)
        trans.transform.translation = t.position
        trans.transform.rotation = t.orientation
        trans.header.frame_id = "base_link"
        trans.header.stamp = rospy.Time.now()
        trans.child_frame_id = "arm_end_effector"

        self.pose_broadcaster.sendTransform(trans)
        return trans

    def get_jacobian(self, joint):
        jac = PyKDL.Jacobian(self.kine_chain.getNrOfJoints())
        self.jac_calc.JntToJac(joint, jac)
        return jac

    def inverse_kinematics_closed(self, desired_pose):

        required_joint = PyKDL.JntArray(self.kine_chain.getNrOfJoints())
        # Current joint array, desired frame, required joint
        self.ik_solver.CartToJnt(self.current_joint_position, desired_pose, required_joint)

        return required_joint